From 9e38a126e87de899ce4cb8232fc13671c546b3af Mon Sep 17 00:00:00 2001
From: Michal Schmidt <mschmidt@redhat.com>
Date: Fri, 5 Sep 2014 19:49:33 -0400
Subject: [ethernet] bnx2x: Fix possible memory leak on iov error flow

Message-id: <1409946615-5862-29-git-send-email-mschmidt@redhat.com>
Patchwork-id: 91573
O-Subject: [RHEL7.1 PATCH 28/70] bnx2x: Fix possible memory leak on iov error flow
Bugzilla: 1089402
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1089402

Upstream commit:
commit dae98cadf743daa480c357b7ba4410423c5f905f
Author: Yuval Mintz <Yuval.Mintz@qlogic.com>
Date:   Thu Mar 27 13:46:37 2014 +0200

    bnx2x: Fix possible memory leak on iov error flow

    Commit 2dc33bbc4 "bnx2x: Remove the sriov VFOP mechanism" introduced a possible
    memory leak on the error flow during multicast filters configuration.

    Reported-by: Dan Carpenter <dan.carpenter@oracle.com>
    Signed-off-by: Yuval Mintz <Yuval.Mintz@qlogic.com>
    Signed-off-by: Ariel Elior <Ariel.Elior@qlogic.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_sriov.c b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_sriov.c
index addb79c..0399227 100644
--- a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_sriov.c
+++ b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_sriov.c
@@ -597,6 +597,8 @@ int bnx2x_vf_mcast(struct bnx2x *bp, struct bnx2x_virtf *vf,
  rc = bnx2x_config_mcast(bp, &mcast, BNX2X_MCAST_CMD_DEL);
  if (rc) {
   BNX2X_ERR("Failed to remove multicasts\n");
+  if (mc)
+   kfree(mc);
   return rc;
  }
 
-- 
1.7.1